### ISSUE TITLE
[Bug] FSDP1 block statistics are not reduced when each rank sees only one GPU

### ISSUE THREAD

**Issue by gss10282023** [zyushun/Adam-mini#42]
### Summary

Adam-mini uses `torch.cuda.device_count()` as `self.world_size`. In a two-rank job where each process sees only its own GPU, this is 1 on both ranks even though the process group has two members.

For a default scalar block that FSDP1 has split across ranks, the optimizer then skips the block-statistic all-reduce. Each shard computes its own second-moment mean instead of the mean over the original block. Nothing raises; training just continues with the wrong statistics.

### Environment

Commit [402aadfe](https://github.com/zyushun/Adam-mini/blob/402aadfed23719f3b92c8e78b5d49ad499ff7ada/adam_mini/adam_mini.py), PyTorch 2.7.0+cu128, real FSDP1 with `use_orig_params=True`, two RTX 5090 ranks with one visible GPU each, FP64.

### Steps to reproduce

Use one `norm.weight` block with four elements and global gradient `[1, 1, 3, 3]`. Compare `NO_SHARD` and `FULL_SHARD` from the same initial state with `beta2=0.99`.

After the first update the block statistic should be:

```text
vmean = (1 - 0.99) * mean([1, 1, 9, 9]) = 0.05
```

| Layout | First-step block statistic |
| --- | --- |
| `NO_SHARD` | `0.05` |
| `FULL_SHARD`, rank 0 | `0.01` |
| `FULL_SHARD`, rank 1 | `0.09` |

On both sharded ranks `torch.cuda.device_count() == 1` while `dist.get_world_size() == 2`. The gathered global gradients match the unsharded run, so the input is the same; only the reduction is missing.

Over three steps the max parameter difference between the two layouts is `0.1554894351`. Reducing over the block's actual process group brings it to zero and matches an independent calculation of the Adam-mini update. Reproduced on a second two-GPU host.

### Expected behavior

Block statistics should be reduced across the processes that own shards of that block. The number of GPUs visible to one process says nothing about which processes those are.

### Root cause

The guard is in the default-block path of `adam_mini/adam_mini.py`: the all-reduce only runs when `self.world_size > 1`, and `world_size` comes from `device_count()`.

[#35](https://github.com/zyushun/Adam-mini/issues/35) is the mirror image of this: several visible GPUs, so the optimizer tries to communicate without an initialized process group. Here the process group is initialized and the optimizer silently skips communication.

### Proposed fix

Take the sharding information explicitly rather than inferring it from visible devices. Switching every optimizer to the default process group would be wrong for other layouts, so the fix below is opt-in and limited to plain-Tensor default scalar blocks. DTensor, head/neuron blocks and hybrid parallelism are not covered.
